From 9f79f8e02289d6ce3b485d1fb3064323abbe69eb Mon Sep 17 00:00:00 2001
From: Dong Chuanxiao <chuanxiao.dong@intel.com>
Date: Wed, 17 Jul 2013 15:13:05 +0800
Subject: [PATCH 236/429] mmc: init eMMC card in sync mode

Since adb device ID is depends on eMMC CID, let MMC driver init
eMMC card in sync mode can make sure eMMC CID be got by USB driver
correctly.
In the meanwhile, it is also can make sure before IPC send command
to SCU to access eMMC card, eMMC card is initliazed already.

Signed-off-by: Ning Feiyi <feiyix.ning@intel.com>
Signed-off-by: Chuanxiao Dong <chuanxiao.dong@intel.com>
---
 drivers/mmc/core/core.c      |  2 ++
 drivers/mmc/host/sdhci-pci.c |  1 +
 include/linux/mmc/host.h     | 10 ----------
 3 files changed, 3 insertions(+), 10 deletions(-)

diff --git a/drivers/mmc/core/core.c b/drivers/mmc/core/core.c
index bb08d286162d..0fb3b28c653c 100644
--- a/drivers/mmc/core/core.c
+++ b/drivers/mmc/core/core.c
@@ -2429,6 +2429,8 @@ void mmc_start_host(struct mmc_host *host)
 	else
 		mmc_power_up(host);
 	mmc_detect_change(host, 0);
+	if (host->caps2 & MMC_CAP2_INIT_CARD_SYNC)
+		flush_work_sync(&host->detect.work);
 }
 
 void mmc_stop_host(struct mmc_host *host)
diff --git a/drivers/mmc/host/sdhci-pci.c b/drivers/mmc/host/sdhci-pci.c
index af2e8313f2b2..1b7bbbad78cc 100644
--- a/drivers/mmc/host/sdhci-pci.c
+++ b/drivers/mmc/host/sdhci-pci.c
@@ -335,6 +335,7 @@ static int mfd_emmc_probe_slot(struct sdhci_pci_slot *slot)
 	case PCI_DEVICE_ID_INTEL_MFD_EMMC0:
 		mfd_emmc_mutex_register(slot);
 		sdhci_alloc_panic_host(slot->host);
+		slot->host->mmc->caps2 |= MMC_CAP2_INIT_CARD_SYNC;
 		break;
 	case PCI_DEVICE_ID_INTEL_MFD_EMMC1:
 		break;
diff --git a/include/linux/mmc/host.h b/include/linux/mmc/host.h
index cda37d9feb71..9b5ff719059c 100644
--- a/include/linux/mmc/host.h
+++ b/include/linux/mmc/host.h
@@ -284,16 +284,6 @@ struct mmc_host {
 				 MMC_CAP2_PACKED_WR)
 #define MMC_CAP2_NO_PRESCAN_POWERUP (1 << 14)	/* Don't power up before scan */
 #define MMC_CAP2_INIT_CARD_SYNC	(1 << 15)	/* init card in sync mode */
-#define MMC_CAP2_POLL_R1B_BUSY	(1 << 16)	/* host poll R1B busy*/
-#define MMC_CAP2_RPMBPART_NOACC	(1 << 17)	/* RPMB partition no access */
-#define MMC_CAP2_LED_SUPPORT	(1 << 18)	/* led support */
-#define MMC_CAP2_PWCTRL_POWER	(1 << 19)	/* power control card power */
-#define MMC_CAP2_FIXED_NCRC	(1 << 20)	/* fixed NCRC */
-#define MMC_CAP2_HS200_WA	(1 << 21)	/* WA: 100MHz clock in HS200 */
-#define MMC_CAP2_HS400_1_8V_DDR	(1 << 22)	/* support HS400 */
-#define MMC_CAP2_HS400_1_2V_DDR	(1 << 23)	/* support HS400 */
-#define MMC_CAP2_HS400		(MMC_CAP2_HS400_1_8V_DDR | \
-				MMC_CAP2_HS400_1_2V_DDR)
 
 	mmc_pm_flag_t		pm_caps;	/* supported pm features */
 
-- 
2.37.3

